### PR TITLE
(EDUCATOR-1680) add no content view and tests; upgrade to React 16.2

### PR DESCRIPTION
**Pull request by MichaelRoytman**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -21,10 +21,10 @@
     "js-cookie": "^2.1.4",
     "popper.js": "^1.12.5",
     "prop-types": "^15.5.10",
-    "react": "^16.1.0",
+    "react": "^16.2.0",
     "react-dom": "^16.1.0",
-    "react-paginate": "^5.0.0",
     "react-dropzone": "^4.2.3",
+    "react-paginate": "^5.0.0",
```
Comment (MichaelRoytman) [inline package.json]: I'm not sure what happened here?

File: src/components/AssetsPage/AssetsPage.test.jsx [hunk 2]
```diff
@@ -0,0 +1,160 @@
+import React from 'react';
+import Enzyme from 'enzyme';
+import { AssetsPage } from './index';
+
+// import { assetActions } from '../../data/constants/actionTypes';
```
Comment (MichaelRoytman) [inline src/components/AssetsPage/AssetsPage.test.jsx]: TODO: remove.

File: src/components/AssetsPage/index.jsx [hunk 3]
```diff
@@ -10,43 +10,98 @@ import { getAssets } from '../../data/actions/assets';
 import edxBootstrap from '../../SFE.scss';
 import styles from './AssetsPage.scss';
 
-class AssetsPage extends React.Component {
+export class AssetsPage extends React.Component {
```
Comment (MichaelRoytman) [inline src/components/AssetsPage/index.jsx]: I've exported the unconnected component to be able to test it independent of the Redux store.

File: src/components/AssetsPage/index.jsx [hunk 4]
```diff
@@ -10,43 +10,98 @@ import { getAssets } from '../../data/actions/assets';
...
   constructor(props) {
     super(props);
+
+    const hasAssets = this.hasAssets(props.assetsList);
```
Comment (MichaelRoytman) [inline src/components/AssetsPage/index.jsx]: I'm keeping track whether the page has assets or not in the application state.

File: src/components/AssetsPage/index.jsx [hunk 5]
```diff
@@ -10,43 +10,98 @@ import { getAssets } from '../../data/actions/assets';
...
     };
   }
 
   componentDidMount() {
     this.props.getAssets(this.props.request, this.props.courseDetails);
   }
 
-  componentDidUpdate() {
-    this.props.getAssets(this.props.request, this.props.courseDetails);
+  componentWillReceiveProps(nextProps) {
```
Comment (MichaelRoytman) [inline src/components/AssetsPage/index.jsx]: This handles changing hasAssets appropriately, based on the assetsList state. If the page hasAssets, and the nextProps have no assets, we switch to having no assets. If the page hasNoAssets, and the nextProps have assets, we switch to having assets.

assets + no content => no content
no content + new assets => new assets

File: src/components/AssetsPage/index.jsx [hunk 6]
```diff
@@ -10,43 +10,98 @@ import { getAssets } from '../../data/actions/assets';
...
   }
 
+  renderAssetsDropZone = () => (
+    <WrappedAssetsDropZone
+      maxFileSizeMB={this.props.uploadSettings.max_file_size_in_mbs}
+    />
+  )
+
+  renderAssetsPage = () => (
+    <React.Fragment>
```
Comment (MichaelRoytman) [inline src/components/AssetsPage/index.jsx]: This is some new React 16.2 functionality. It basically allows us to return adjacent JSX elements without wrapping them in, say, a div or a span.

You can read more [here](https://reactjs.org/blog/2017/11/28/react-v16.2.0-fragment-support.html).

File: src/components/AssetsPage/index.jsx [hunk 7]
```diff
@@ -10,43 +10,98 @@ import { getAssets } from '../../data/actions/assets';
...
     });
   }
 
+  renderAssetsDropZone = () => (
+    <WrappedAssetsDropZone
+      maxFileSizeMB={this.props.uploadSettings.max_file_size_in_mbs}
+    />
+  )
+
+  renderAssetsPage = () => (
```
Comment (MichaelRoytman) [inline src/components/AssetsPage/index.jsx]: This renders the assets page with content.

File: src/components/AssetsPage/index.jsx [hunk 8]
```diff
@@ -10,43 +10,98 @@ import { getAssets } from '../../data/actions/assets';
...
+        <WrappedAssetsFilters />
+      </div>
+      <div className={edxBootstrap['col-10']}>
+        <WrappedAssetsTable />
+        <WrappedPagination />
+      </div>
+    </React.Fragment>
+  )
+
+  renderNoAssetsPage = () => (
```
Comment (MichaelRoytman) [inline src/components/AssetsPage/index.jsx]: This renders the assets page without content.

File: src/components/AssetsPage/index.jsx [hunk 9]
```diff
@@ -10,43 +10,98 @@ import { getAssets } from '../../data/actions/assets';
...
+      <div className={edxBootstrap.col}>
+        { this.renderAssetsDropZone() }
+      </div>
+      <div className={edxBootstrap['col-10']}>
+        { !this.state.hasAssets && this.renderNoAssetsBody() }
+      </div>
+    </React.Fragment>
+  )
+
+  renderNoAssetsBody = () => (
```
Comment (MichaelRoytman) [inline src/components/AssetsPage/index.jsx]: This is the no content view body.

**Comment by MichaelRoytman** [comment]
If you want to see this no content view in your browser, you need a course that has no assets in it. Instead of deleting all the assets in the demo course, you can just create a new dummy course in studio. Then, change the id [here](https://github.com/edx/studio-frontend/blob/master/public/index.html#L21) to match the course id for your course, which you can grab from the URL when you're in your course.

File: package-lock.json [hunk 10]
```diff
@@ -1582,7 +1582,7 @@
     "browserslist": {
       "version": "2.11.0",
       "resolved": "https://registry.npmjs.org/browserslist/-/browserslist-2.11.0.tgz",
-      "integrity": "sha512-mNYp0RNeu1xueGuJFSXkU+K0nH+dBE/gcjtyhtNKfU8hwdrVIfoA7i5iFSjOmzkGdL2QaO7YX9ExiVPE7AY9JA==",
+      "integrity": "sha1-UDUNaHOoLr4POuVINljFca5fnX0=",
```
Comment (efischer19) [inline package-lock.json]: I'm a little concerned about the size of the package-lock diff here. Why did all these `integrity` hashes update from sha512 to sha1? Are we going to have this amount of thrash with every PR?

File: src/components/AssetsPage/index.jsx [hunk 11]
```diff
@@ -10,43 +10,98 @@ import { getAssets } from '../../data/actions/assets';
...
+        hasAssets: false,
+      });
+    } else if (!this.state.hasAssets && this.hasAssets(nextProps.assetsList)) {
+      this.setState({
+        hasAssets: true,
+      });
+    }
+  }
+
+  componentDidUpdate(prevProps) {
```
[the remark this answers is not in the document]
Comment (efischer19) [reply src/components/AssetsPage/index.jsx]: 👍 from me - "We need a deep equality check b/c the default shallow comparison isn't working" feels like a reasonable reason to violate the "let's not use JSON.stringify" guideline

File: src/components/AssetsPage/index.jsx [hunk 12]
```diff
@@ -10,43 +10,97 @@ import { getAssets } from '../../data/actions/assets';
...
+      <div className={edxBootstrap['col-10']}>
+        { !this.state.hasAssets && this.renderNoAssetsBody() }
+      </div>
+    </React.Fragment>
+  )
+
+  renderNoAssetsBody = () => (
+    <React.Fragment>
+      <h3>0 files in your course</h3>
+      <h4>Enhance your course content by uploading files such as images and documents.</h4>
```
Comment (efischer19) [inline src/components/AssetsPage/index.jsx]: This display text will need to be extracted for i18n purposes, flagging for myself once this gets combined with https://github.com/edx/studio-frontend/pull/84.

[example of a similar transformation](https://github.com/edx/studio-frontend/pull/84/files#diff-e5d111258070606d64e86dc6dbcd0a42L56)

File: package.json (hunk 1 above)
Comment (fysheets) [reply package.json]: Looks like it alphabetized?

**Pull request merged**